From a37dc8f3f952f07cb40676cb55093435a2fb7495 Mon Sep 17 00:00:00 2001
From: "Liu,Dalei" <daleix.liu@intel.com>
Date: Mon, 18 May 2015 14:55:10 +0800
Subject: [PATCH 402/462] dw9718: Add power down sequence

This patch add dw9718 power down sequence.
Sometimes dw9718 didn't power cycle because v2p8 was still used by
someone. So we need to reset dw9718 even its power rail was not off.

Change-Id: Id8dc6d2914ca7f672106bbbb0bba577195335200
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-33416
Signed-off-by: Liu,Dalei <daleix.liu@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/56347
Reviewed-by: Jian Xu Zheng <jian.xu.zheng@intel.com>
Reviewed-by: Soren Friis <soren.friis@intel.com>
Build: IMIN (SC)
Validation-Android: Wei Tao <wei.tao@intel.com>
Approver: Fang Wang <fang.c.wang@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Approver: Jing Tao <jing.tao@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c b/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c
index e20b65e..225b477 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c
@@ -108,6 +108,9 @@ int dw9718_vcm_power_up(struct v4l2_subdev *sd)
 	ret = dw9718_dev.platform_data->power_ctrl(sd, 1);
 	if (ret)
 		return ret;
+	ret = dw9718_i2c_wr8(client, DW9718_PD, 0);
+	if (ret < 0)
+		goto fail_powerdown;
 	/* Wait t_OPR for VBAT to stabilize */
 	usleep_range(100, 110);
 
@@ -189,6 +192,11 @@ int dw9718_vcm_power_down(struct v4l2_subdev *sd)
 		i -= step;
 	}
 
+	ret = dw9718_i2c_wr8(client, DW9718_PD, 1);
+	if (ret)
+		dev_err(&client->dev, "%s: write DW9718_PD to 1 failed\n",
+				__func__);
+
 	return dw9718_dev.platform_data->power_ctrl(sd, 0);
 }
 
-- 
1.9.1

